binstat/binstat.go

```diff
@@ -0,0 +1,530 @@
+package binstat
+
+/*
+#if defined(__linux__)
+#include <stdint.h>
+#include <unistd.h>
+#include <sys/syscall.h>
+#ifdef SYS_gettid
+uint64_t gettid() { return syscall(SYS_gettid); }
+#else
+#error "SYS_gettid unavailable on this system"
+#endif
+#elif __APPLE__ // http://elliotth.blogspot.com/2012/04/gettid-on-mac-os.html
+#include <stdint.h>
+#include <pthread.h>
+uint64_t gettid() { uint64_t tid; pthread_threadid_np(NULL, &tid); return tid; }
+#else
+#   error "Unknown platform; __linux__ or __APPLE__ supported"
+#endif
+*/
+import "C"
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"runtime/debug"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/rs/zerolog"
+)
+
+// todo: consider using something faster than runtime.nano(), eg. https://github.com/templexxx/tsc/issues/8 <-- but maybe not safe with multiple CPUs? https://stackoverflow.com/questions/3388134/rdtsc-accuracy-across-cpu-cores
+// "Another concern is that if a thread is migrated on a different processor between two measurements, the counter might skip too much or even “go back”. " https://coherent-labs.com/posts/timestamps-for-performance-measurements/
```

Comment:
Consider wrapping lines, long lines are hard to read

Reply:
There is a reason to have the longer 'todo:' lines: So that it's easier to grep them out from the command line if necessary.

Reply:
But upon further examination, in this case it seems like the todo can be refactored shorter! :-)